### PR TITLE
140/optimization extension

### PR DESCRIPTION
**Pull request by carlaterboven**

### CODE REVIEW

File: sphinx/optimization/optimizationStrategy.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    feasibility(stop1, parcel, stop2) { }
+
+    feasibilityLastStop(stop, parcel, profit, tourlength) { }
+
+    profitCalculation(stop1, stop2) { }
+
+    timeFrameUpdate(stop1, parcel, stop2) { }
+
+    timeFrameUpdateLastStop(stop, parcel, profit) { }
```
Comment (svenihde) [inline sphinx/optimization/optimizationStrategy.js]: These methods above have not functionality yet. Is that planned or are they no longer needed?

File: sphinx/optimization/travelTimeStrategy.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            let newTimeFrameEnd = new Date(Math.min(this.tours[routesIndexUpdate][i - 1].plannedTimeFrameEnd.getTime(), timeCalculation));
+
+            //only need to continue the updates as long as the value for the most recently evaluated customer changed
+            if (newTimeFrameEnd === this.tours[routesIndexUpdate][i - 1].plannedTimeFrameEnd) {
+                break;
+            } else {
+                this.tours[routesIndexUpdate][i - 1].plannedTimeFrameEnd = newTimeFrameEnd;
+            }
+        }
+        //TODO warum geht <= nicht mehr?
```
Comment (JonasZim) [inline sphinx/optimization/travelTimeStrategy.js]: is this still a TODO

File: sphinx/routes/index.js [hunk 3]
```diff
@@ -81,6 +81,41 @@ router.get('/deleteTours', function (req, res, next) {
...
 });
 
+const jsonToInput = (jsonAddresses) => {
+    let addresses = [];
+    jsonAddresses.forEach((jsonAddress) => {
+        addresses.push({
+            streetName: jsonAddress.streetName,
+            streetNumber: jsonAddress.streetNumber,
+            city: jsonAddress.city,
+            wtf: {start: new Date(jsonAddress.wtf.start), end: new Date(jsonAddress.wtf.end)}
```
Comment (JonasZim) [inline sphinx/routes/index.js]: is wtf.end necessary?

Comment (carlaterboven) [reply sphinx/routes/index.js]: Yes otherwise you wouldn't know the closing time of the depot

File: sphinx/optimization/travelTimeStrategy.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        let timeFrameStart = Math.max(parcel.wtf.start, stop1.plannedTimeFrameStart.getTime() + this.profitCalculation(stop1, parcel));
+        let timeFrameEnd = Math.min(parcel.wtf.end, stop2.plannedTimeFrameEnd.getTime() - this.profitCalculation(parcel, stop2));
+
+        return (timeFrameStart <= timeFrameEnd);
+    }
+
+    profitCalculation(stop1, stop2) {
+        //+60 for the time the deliverer needs for one stop (1min)
+        //*1000 to convert seconds into milliseconds
+        return (this.matrix[stop1.optimizationIndex][stop2.optimizationIndex] + 60) * 1000;
```
Comment (svenihde) [inline sphinx/optimization/travelTimeStrategy.js]: At this point does the matrix has time as the measurement between points/stops or distance?

File: sphinx/optimization/travelTimeStrategy.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    timeFramePreparation(address) {
+        address.plannedTimeFrameStart = address.wtf.start;     //already exists in database
+        address.plannedTimeFrameEnd = address.wtf.end;
+    }
+
+    updateRoute(routesIndexUpdate, tourStopIndexUpdate) {
+        //update plannedTimeFrameStart and plannedTimeFrameEnd in the whole tour
+        for (let i = tourStopIndexUpdate; i > 0; i--) {
+            let timeCalculation = this.tours[routesIndexUpdate][i].plannedTimeFrameEnd.getTime() - this.profitCalculation(this.tours[routesIndexUpdate][i - 1], this.tours[routesIndexUpdate][i]);
```
Comment (svenihde) [inline sphinx/optimization/travelTimeStrategy.js]: Good idea to put a comment here to describe what time calculation is supposed to be used for. Example: "Time Calculation finds the last possible time of the previous stop to finish in order to be in time for this stop"

**Pull request merged**